default/python/universe_generation/empires.py

```diff
@@ -95,26 +85,80 @@ def min_planets_in_vicinity_limit(num_systems):
-        # add the result of the current attempt to our list and continue with next attempt
-        accepted_list.append(accepted)
-        attempts -= 1
-    # return the result of the attempt that yielded the most systems
-    return max(accepted_list, key=len)
+class HomeSystemFinder(object):
+    """Finds a set of home systems with a least ''num_home_systems'' systems."""
+    def __init__(self, _num_home_systems):
+        # cache of sytem merits
+        self.system_merit = {}
+        self.num_home_systems = _num_home_systems
+
+    def find_home_systems_for_min_jump_distance(self, systems_pool, min_jumps):
+        """
+        Return a good list of home systems or an empty list if there are fewer than num_home_systems in the pool.
+
+        A good list of home systems are at least the specified minimum number of jumps apart,
+        with the best minimum system merit of all such lists picked randomly from the ''systems_pool''.
+
+        Algorithm:
+        Make several attempts to find systems that match the condition
+        of being at least min_jumps apart.
+        Use the minimum merit of the best num_home_system systems found
+        as the merit of the current best set of systems.
+        On each attempt use the merit of the current best set of home
+        systems to truncate the pool of candidates.
+        """
+
+        # precalculate the system merits
+        for system in systems_pool:
+            if system not in self.system_merit:
+                self.system_merit[system] = calculate_home_system_merit(system)
+
+        # In the list of systems sorted by merit, if the first num_home_systems are all min_jumps
+        # apart, then the merit of the num_home_systems nth system is the
+        # best_case_merit_lower_bound that can be achieved.
+        best_case_merit_lower_bound = sorted(self.system_merit.values())[self.num_home_systems - 1]
+        current_merit_lower_bound = 0
+
+        best_candidate = []
+
+        # cap the number of attempts to the smaller of the number of systems in the pool, or 100
+        attempts = min(100, len(systems_pool))
+
+        # Copy systems_pool to avoid aliasing of the caller's systems_pool variable
+        local_pool = set(systems_pool)
+
+        while attempts:
+            candidate = []
+            # use a local pool of all candidate systems better than the worst threshold merit
+            local_pool = {s for s in local_pool if self.system_merit[s] > current_merit_lower_bound}
+            if len(local_pool) < self.num_home_systems:
+                break
+
+            while local_pool:
+                member = random.choice(list(local_pool))
+                candidate.append(member)
+
+                # remove all neighbors from the local pool
+                local_pool -= set(get_systems_within_jumps(member, min_jumps))
+
+            if len(candidate) < self.num_home_systems:
+                break
```

Comment:
Why breaking out of the loop here, instead of decreasing attempts and `continue`? The reason for coming up with too few candidates might have been a very unlucky selection, so subsequent attempts might be successful.